_posts/2023/03/2023-03-13-probing.md

```diff
@@ -0,0 +1,178 @@
++probe org.eolang # Not found -
++probe org.eolang.io # Not found -
++probe org.eolang.txt # Not found -
++probe org.eolang.io.stdout # Found +
++probe org.eolang.txt.sprintf # Found +
+```
+
+In the example mentioned earlier, the probing process will only locate two
+top-level objects in the Objectionary: `org.eolang.io.stdout` and
+`org.eolang.txt.sprintf`.
+
+## Optimizations
+
+The naive implementation of checking each object in the Objectionary one by one
+is obviously not the most efficient in terms of speed and effectiveness. To
+speed up the probing process, several improvements have been implemented. These
+improvements make the probing process more efficient and accurate, while also
+reducing the overall execution time of the compiler.
+
+### Local objects
+
+The first and simplest optimization to improve the probing process is to use
+local files and cache. This means that not all objects need to be found in the
+Objectionary. Some objects may be present locally in the program text or in the
+local cache. The compiler will first check these local sources before searching
+for the objects in the Objectionary. By utilizing local files and cache, the
+probing process becomes faster and more efficient, as the compiler does not need
+to spend time searching for objects that are already present locally.
+
+### Filtering
+
+Another optimization that was implemented is filtering out certain objects
+that are not relevant to the probing process. For example, in the Objectionary,
+objects with a package length of less than 3 do not exist, which
+means that objects such as `org`, `com`, and `org.eolang` are not valid for
+probing. These objects can be filtered out during the probing process, reducing
+the number of unnecessary checks and making the process faster and more
+efficient. By filtering out irrelevant objects, the compiler can focus on
+locating the objects that are actually necessary for the program.
+
+### Index file
+
+Checking each object in the Objectionary one by one can also result in a high
+number of HTTP requests, which is not ideal in terms of performance. To improve
+the speed of the probing process, an index file has been added that contains a
+list of all objects in the Objectionary. This index file is downloaded once, and
+then used to decrease the overall number of check requests required during the
+probing process. The index file can be accessed at
+this [link](https://github.com/objectionary/home/blob/gh-pages/objectionary.lst),
+and its content appears as follows:
+
+```txt
+objects/org/eolang/bytes.eo
+objects/org/eolang/string.eo
+objects/org/eolang/seq.eo
+objects/org/eolang/as-phi.eo
+objects/org/eolang/rust.eo
+objects/org/eolang/int.eo
+objects/org/eolang/txt/regex.eo
+objects/org/eolang/txt/text.eo
+objects/org/eolang/txt/sprintf.eo
+...
+objects/org/eolang/io/stdout.eo
+...
+objects/org/eolang/txt/sprintf.eo
+...
+```
+
+The index file contains a list of top-level objects that can be used in the
```

Comment:
@volodya-lombrozo again, it's not fully correct to use naming "top-level objects". We have no such notion.

Reply:
@mximp Let's discuss it [here](https://github.com/objectionary/news.eolang.org/pull/29#discussion_r1134179269).